README.md
# CosmicRay

Basic tools for cosmic ray visualization

# How to use

## Setup File paths and modules

1. Clone this repo locally
2. Run the file *RUN_ME_FIRST.py*
3. Check to make sure that directories *DataDates*, *Movies*, and *RawData* have been created.
4. run: *pip install -r requirments.txt*

##  Get data

1. Log onto tadserv
2. Move into */scratch0/rasha/lv10rate/outFiles*
3. Copy contents locally, into *CosmicRay/RawData/L0L1*
4. Move into */scratch0/rasha/NLDN_data*
5. Copy *Raw_NLDN.txt* into *CosmicRay/RawData/NLDN*

## Parse dates

We need to extract the data we want from the raw data we have based on the dates we care about. This can be done with the *GetDateData.py* tool
1. Run *GetDateData.py*
2. Choose the date you want
3. Check that the date got a folder within *DataDates* (i.e. *DataDates/140927*)
4. Check that the date got NLDN data and L0L1 data

## Create Movies

1. Run CoreRateLv0.py
2. Choose the date you wish (It has to be parsed beforehand)
3. Let it run - it will output into *Movies/{date}*


src/history.txt

# 2021-10-20 22:46:17.715697
+loadYear help

# 2021-10-20 22:53:46.299836
+cleaar

# 2021-10-20 22:53:53.124060
+loadYear help

# 2021-10-20 23:01:28.822621
+save help

# 2021-10-20 23:03:48.060724
+loadDate

# 2021-10-20 23:04:44.573742
+loadDate 130401

# 2021-10-20 23:06:38.717490
+loadYear

# 2021-10-20 23:06:45.189888
+loadYear 2013

# 2021-10-20 23:18:39.094085
+oneD

# 2021-10-20 23:29:50.831851
+oneD help

# 2021-10-20 23:30:02.319837
+clear

# 2021-10-20 23:30:08.047052
+oneD help

# 2021-10-21 19:55:13.704792
+loadYear help

# 2021-10-21 19:57:10.705099
+oneD 140927 1415 18:27:31 19:27:23

# 2021-10-21 20:05:48.730397
+loadDate

# 2021-10-21 20:05:54.793440
+loadDate 140927

# 2021-10-21 20:05:59.378139
+oneD 140927 1415 18:27:31 19:27:23

# 2021-10-21 20:06:36.449974
+loadDate 140927

# 2021-10-21 20:06:39.618364
+oneD 140927 1415 18:27:31 19:27:23

# 2021-10-21 20:17:12.282731
+loadDate 140927

# 2021-10-21 20:17:15.698842
+oneD 140927 1415 18:27:31 19:27:23

# 2021-10-21 20:17:50.810893
+loadDate 140927

# 2021-10-21 20:17:54.826550
+oneD 140927 1415 18:27:31 19:27:23

# 2021-10-21 20:18:36.906629
+loadDate 140927

# 2021-10-21 20:18:40.667053
+oneD 140927 1415 18:27:31 19:27:23

# 2021-10-21 20:19:04.498805
+loadDate 140927

# 2021-10-21 20:19:08.163487
+oneD 140927 1415 18:27:31 19:27:23

# 2021-10-21 20:19:35.658218
+c

# 2021-10-21 20:19:39.498715
+clear

# 2021-10-21 20:19:47.170710
+loadDate 140927

# 2021-10-21 20:19:52.659213
+oneD 140927 1415 18:27:31 19:27:23

# 2021-10-21 20:20:03.498826
+loadDate 140927

# 2021-10-21 20:20:07.242757
+oneD 140927 1415 18:27:31 19:27:23

# 2021-10-21 20:21:47.923294
+loadDate 140927

# 2021-10-21 20:21:51.699732
+oneD 140927 1415 18:27:31 19:27:23

# 2021-10-21 20:24:11.588112
+loadDate 140927

# 2021-10-21 20:24:14.732525
+oneD 140927 1415 18:27:31 19:27:23

# 2021-10-21 20:24:40.370982
+loadDate 140927

# 2021-10-21 20:24:43.307160
+oneD 140927 1415 18:27:31 19:27:23

# 2021-10-21 20:25:43.972620
+loadDate 140927

# 2021-10-21 20:25:47.283645
+oneD 140927 1415 18:27:31 19:27:23

# 2021-10-21 20:31:28.139676
+loadDate 140927

# 2021-10-21 20:31:31.595823
+oneD 140927 1415 18:27:31 19:27:23

# 2021-10-21 20:32:42.020240
+cle

# 2021-10-21 20:32:48.003689
+loadDate 140927

# 2021-10-21 20:32:51.148064
+oneD 140927 1415 18:27:31 19:27:23

# 2021-10-21 20:35:51.395895
+loadDate 140927

# 2021-10-21 20:35:54.900601
+oneD 140927 1415 18:27:31 19:27:23

# 2021-10-21 20:37:32.874044
+loadDate 140927

# 2021-10-21 20:37:35.644082
+oneD 140927 1415 18:27:31 19:27:23

# 2021-10-21 20:39:15.653005
+loadDate 140927

# 2021-10-21 20:39:25.197695
+oneD 140927 1415 18:27:31 19:27:23

# 2021-10-21 20:42:11.932439
+loadDate 140927

# 2021-10-21 20:42:33.588292
+oneD 140927 1415 18:27:31 19:27:23

# 2021-10-21 20:42:43.948256
+oneD 140927 1415 18:20:00 19:20:00

# 2021-10-21 20:44:11.893045
+loadDate 140927

# 2021-10-21 20:44:16.644371
+oneD 140927 1415 18:20:00 19:20:00

# 2021-10-21 20:48:29.789667
+loadDate 140927

# 2021-10-21 20:48:36.853612
+oneD 140927 1415 18:20:00 19:20:00

# 2021-10-21 20:50:54.645244
+loadDate 140927

# 2021-10-21 20:50:58.021345
+oneD 140927 1415 18:20:00 19:20:00

# 2021-10-21 20:53:32.422019
+loadDate 140927

# 2021-10-21 20:54:27.269494
+oneD 140927 2220 23:50:00 23:50:00

# 2021-10-21 20:55:05.974931
+loadDate 140927

# 2021-10-21 20:55:09.486151
+oneD 140927 2220 23:50:00 23:50:00

# 2021-10-21 20:56:01.038209
+loadDate 140927

# 2021-10-21 20:56:04.182099
+oneD 140927 2220 23:50:00 23:50:00

# 2021-10-21 21:22:37.073182
+loadDate 140927

# 2021-10-21 21:22:40.391709
+oneD 140927 2220 23:50:00 23:50:00

# 2021-10-21 21:26:28.199268
+loadDate 140927

# 2021-10-21 21:26:31.434729
+oneD 140927 2220 23:50:00 23:50:00

# 2021-10-21 21:28:17.969343
+loadDate 140927

# 2021-10-21 21:28:21.440504
+oneD 140927 2220 23:50:00 23:50:00

# 2021-10-21 21:28:40.376270
+loadDate 140927

# 2021-10-21 21:28:43.816908
+oneD 140927 2220 23:50:00 23:50:00

# 2021-10-21 21:29:10.257514
+loadDate 140927

# 2021-10-21 21:29:13.552523
+oneD 140927 2220 23:50:00 23:50:00

# 2021-10-21 21:30:04.784414
+loadDate 140927

# 2021-10-21 21:30:08.536841
+oneD 140927 2220 23:50:00 23:50:00

# 2021-10-21 21:39:30.913425
+loadDate 140927

# 2021-10-21 21:39:34.569267
+oneD 140927 2220 23:50:00 23:50:00

# 2021-10-21 21:39:55.317711
+loadDate 140927

# 2021-10-21 21:39:58.569682
+oneD 140927 2220 23:50:00 23:50:00

# 2021-10-21 21:40:14.951618
+loadDate 140927

# 2021-10-21 21:40:17.946345
+oneD 140927 2220 23:50:00 23:50:00

# 2021-10-21 21:40:28.377371
+loadDate 140927

# 2021-10-21 21:40:31.569781
+oneD 140927 2220 23:50:00 23:50:00

# 2021-10-21 21:41:19.762294
+loadDate 140927

# 2021-10-21 21:41:23.074814
+oneD 140927 2220 23:50:00 23:50:00

# 2021-10-21 21:41:37.057557
+loadDate 140927

# 2021-10-21 21:41:40.553768
+oneD 140927 2220 23:50:00 23:50:00

# 2021-10-21 21:42:04.387066
+loadDate 140927

# 2021-10-21 21:42:07.778167
+oneD 140927 2220 23:50:00 23:50:00

# 2021-10-21 21:42:35.993968
+loadDate 140927

# 2021-10-21 21:42:39.506176
+oneD 140927 2220 23:50:00 23:50:00

# 2021-10-21 21:42:56.714511
+loadDate 140927

# 2021-10-21 21:43:03.249966
+oneD 140927 2220 23:40:00 23:50:00

# 2021-10-21 21:43:13.545883
+loadDate 140927

# 2021-10-21 21:43:35.481877
+oneD 140927 1515 09:10:00 10:10:00

# 2021-10-21 21:44:26.258189
+loadDate 140927

# 2021-10-21 21:44:29.450616
+oneD 140927 1515 09:10:00 10:10:00

# 2021-10-21 21:46:16.770017
+loadDate 140927

# 2021-10-21 21:46:20.042023
+oneD 140927 1515 09:10:00 10:10:00

# 2021-10-21 21:46:55.857738
+loadDate 140927

# 2021-10-21 21:47:01.274199
+oneD 140927 1515 09:10:00 10:10:00

# 2021-10-21 21:48:34.217875
+oneD 140927 1515 09:10:00 10:10:00 10:00:00

# 2021-10-21 21:49:51.666002
+oneD 140927 1515 09:10:00 10:10:00

# 2021-10-21 21:49:53.922056
+loadDate 140927

# 2021-10-21 21:49:57.170518
+oneD 140927 1515 09:10:00 10:10:00

# 2021-10-21 21:50:19.394315
+loadDate 140927


CRAT.py
import os
import sys
from typing import List

from prompt_toolkit import prompt
from prompt_toolkit.auto_suggest import AutoSuggestFromHistory
from prompt_toolkit.completion import WordCompleter
from prompt_toolkit.completion.word_completer import WordCompleter
from prompt_toolkit.history import FileHistory

from src.dataDates import DataDates
from src.OneDimension import getDataForOneD, plotOneD
from src.Utils import stringToDatetime

class DataDatePrompt:
    
    def __init__(self) -> None:

        self._dates = DataDates()

        self._commands = {
            'help': self._help,
            'exit': '_',
            'clear': self._clear,
            'loadYear': self._dates.loadAndSaveYear,
            'loadDate': self._dates.newDataDate,
            'save': self._save,
            'oneD': self._oneD,
        }

        self._commands_help = {
            'exit': self._des_exit,
            'clear': self._des_clear,
            'loadYear': self._des_loadYear,
            'loadDate': self._des_loadDate,
            'save': self._des_save,
            'oneD': self._des_oneD,
        }

        command_completer = WordCompleter(self._commands.keys())

        while True:
            user_input = prompt(u'/> ',
                                history=FileHistory('history.txt'),
                                auto_suggest=AutoSuggestFromHistory(),
                                ).split()

            user_command, args = user_input[0], user_input[1:]

            try:
                if 'help' in args:
                    self._commands_help[user_command]()
                elif user_command == 'exit':
                    break
                elif user_command in self._commands:
                    self._commands[user_command](*args)
                else:
                    print('Not a command!')
            except TypeError as e:
                    print(str(e).split(' ', 1)[1])

    def _des_exit(self) -> None:
        self._des('exit', 'Will exit the terminal', ['exit'])

    def _des_clear(self) -> None:
        self._des('clear', 'Will clear the screen', ['clear'])

    def _des_loadYear(self) -> None:
        self._des('loadYear', 'Will load a years worth of data into buffer based on any NLDN activity and save to files (NOT TO MOVIES)', ['loadYear 2015'], ['year'])

    def _des_loadDate(self) -> None:
        self._des('loadDate', 'Will load a specified day into buffer', ['loadDate 120121'], ['date'])

    def _des_save(self) -> None:
        self._des('save', 'Will save buffered dates into files or movies', ['save dates', 'save movies'], ['"dates" || "movies"'])
    
    def _des_oneD(self) -> None:
        self._des('oneD', 'Will create a one dimension level0 and tempature plot from a LOADED date', ['oneD 140927 1415 18:20:00 19:20:00', 'oneD 150515 1308 07:00:00 16:30:00 8:10:00'], ['date', 'detector number', 'starting time', 'ending time', '~marked time'])

    def _des(self, name: str, des: str, examples: List[str], func_inputs: List[str] = None) -> None:
        print(f'\n{name}\n')
        print(f'     {des}\n')

        if func_inputs:
            print('     Inputs: ', end='')
            for func_input in func_inputs:
                print(f'({func_input}) ', end='')
            print('\n')

        print(f'     Examples: {examples[0]}')
        for example in examples[1:]:
            print(f'               {example}')
        print()

    def _help(self) -> None:
        for help_command in self._commands_help:
            print('========================================================================================')
            self._commands_help[help_command]()
            print('========================================================================================')

    def _clear(self) -> None:
        if sys.platform == 'linux' or sys.platform == 'linux2' or sys.platform == 'darwin':
            os.system('clear')
        elif sys.platform == 'win32':
            os.system('cls')

    def _save(self, arg: str) -> None:

        valid_args = {
            'dates': self._dates.saveDates,
            'movies': self._dates.saveMovies
        }

        if arg not in valid_args.keys():
            print('\nNot a valid argument')

        else:
            valid_args[arg]()

    def _oneD(self, date: str, detector_number: str, start_time: str, end_time: str, marked_time: str = None) -> None:
        if date in self._dates.keys():
            one_d_data = getDataForOneD(self._dates[date], detector_number, stringToDatetime(start_time), stringToDatetime(end_time))
            plotOneD(*one_d_data, stringToDatetime(marked_time))
        else:
            print('Date not loaded!')

if __name__ == '__main__':

    try:
        DataDatePrompt()
    except KeyboardInterrupt:
        print('\nTerminal interupted')
